src/passes/MergeSimilarFunctions.cpp

```diff
@@ -0,0 +1,626 @@
+    } else {
+      WASM_UNREACHABLE("unexpected const value type");
+    }
+  }
+
+  // Lower the constant value at a given index to an expression
+  Expression*
+  lowerToExpression(Builder& builder, Module* module, size_t index) const {
+    if (const auto literals = std::get_if<Literals>(&values)) {
+      return builder.makeConst((*literals)[index]);
+    } else if (auto callees = std::get_if<std::vector<Name>>(&values)) {
+      auto fnName = (*callees)[index];
+      auto heapType = module->getFunction(fnName)->type;
+      return builder.makeRefFunc(fnName, heapType);
+    } else {
+      WASM_UNREACHABLE("unexpected const value type");
+    }
+  }
+};
+
+// Describes the set of functions which are considered as "equivalent" (i.e.
+// only differing by some constants).
+struct EquivalentClass {
+  // Primary function in the `functions`, which will be the base for the merged
+  // function.
+  Function* primaryFunction;
+  // List of functions belonging to this equivalence class.
+  std::vector<Function*> functions;
+
+  EquivalentClass(Function* primaryFunction, std::vector<Function*> functions)
+    : primaryFunction(primaryFunction), functions(functions) {}
+
+  bool isEligibleToMerge() { return this->functions.size() >= 2; }
+
+  // Merge the functions in this class.
+  void merge(Module* module, const std::vector<ParamInfo>& params);
+
+  bool hasMergeBenefit(Module* module, const std::vector<ParamInfo>& params);
+
+  Function* createShared(Module* module, const std::vector<ParamInfo>& params);
+
+  Function* replaceWithThunk(Builder& builder,
+                             Function* target,
+                             Function* shared,
+                             const std::vector<ParamInfo>& params,
+                             const std::vector<Expression*>& extraArgs);
+
+  bool deriveParams(Module* module,
+                    std::vector<ParamInfo>& params,
+                    bool isIndirectionEnabled);
+};
+
+struct MergeSimilarFunctions : public Pass {
+  bool invalidatesDWARF() override { return true; }
+
+  void run(PassRunner* runner, Module* module) override {
+    std::vector<EquivalentClass> classes;
+    collectEquivalentClasses(classes, module);
+    std::sort(
+      classes.begin(), classes.end(), [](const auto& left, const auto& right) {
+        return left.primaryFunction->name < right.primaryFunction->name;
+      });
+    for (auto& clazz : classes) {
+      if (!clazz.isEligibleToMerge()) {
+        continue;
+      }
+
+      std::vector<ParamInfo> params;
+      if (!clazz.deriveParams(
+            module, params, isCallIndirectionEnabled(module))) {
+        continue;
+      }
+
+      if (!clazz.hasMergeBenefit(module, params)) {
+        continue;
+      }
+
+      clazz.merge(module, params);
+    }
+  }
+
+  // Parameterize direct calls if the module supports func ref values.
+  bool isCallIndirectionEnabled(Module* module) const {
+    return module->features.hasTypedFunctionReferences();
```

Comment:
```suggestion
    return module->features.hasReferenceTypes();
```

Reply:
(fuzzer noticed this before me...)

Reply:
Oh, actually you use call_ref as well... so you do need both features. Apologies for the noise here.

Reply:
And I think you should check them both, as you had before - again, sorry for my confusion there.

@tlively I guess we don't have a mechanism for making a feature like typed-references automatically enable a feature it depends on like reference types?

Reply:
AFAIU, "TypedFunctionReferences requires ReferenceTypes", and callee parameterization requires TypedFunctionReferences, so I thought `hasTypedFunctionReferences` is appropriate here. If you overlook something, let me know 🙏

Reply:
> @tlively I guess we don't have a mechanism for making a feature like typed-references automatically enable a feature it depends on like reference types?

That's correct, and somewhat deliberate to keep the mental model of enabling and disabling features simpler and more predictable.